From 19bd70a2b113ce1b0ea8fe96b8e4c1b48e4d80dc Mon Sep 17 00:00:00 2001
From: Torsten Veller <tove@gentoo.org>
Date: Sun, 15 Apr 2012 08:56:44 +0200
Subject: [PATCH 02/30] aix gcc detection and shared library soname support

1) Match "powerpc-ibm-aix5.3.0.0-gcc" as gcc too.

2) Creating shared library on aix with full soname support, following
http://lists.gnu.org/archive/html/libtool/2011-01/msg00023.html
http://lists.gnu.org/archive/html/libtool/2011-02/msg00000.html
Once there is an agreement on that, this patch basically is designed
for upstream perl to be applied when their soname would change anyway.
However, there is 'strip' being used, and I don't know yet what to
use instead in cross-aix setups with recent binutils.

Patch-Name: gentoo/aix_soname.diff
---
 Makefile.SH | 32 ++++++++++++++++++++------------
 1 file changed, 20 insertions(+), 12 deletions(-)

diff --git a/Makefile.SH b/Makefile.SH
index 3f1851d0e4..1162b6ed03 100755
--- a/Makefile.SH
+++ b/Makefile.SH
@@ -94,12 +94,12 @@ true)
 		;;
 	aix*)
 		case "$cc" in
-		gcc*)
-			shrpldflags="-shared -Wl,-H512 -Wl,-T512 -Wl,-bhalt:4 -Wl,-bM:SRE -Wl,-bE:perl.exp"
+		*gcc*)
+			shrpldflags="-shared -Wl,-H512 -Wl,-T512 -Wl,-bhalt:4 -Wl,-bE:perl.exp"
 			case "$osvers" in
-			3*)	shrpldflags="$shrpldflags -e _nostart"
+			3*)	shrpldflags="$shrpldflags -Wl,-bM:SRE -e _nostart"
 				;;
-			*)	shrpldflags="$shrpldflags -Wl,-bnoentry"
+			*)	shrpldflags="$shrpldflags -Wl,-G -Wl,-bernotok -Wl,-bnoentry"
 				;;
 			esac
 			shrpldflags="$shrpldflags $ldflags $perllibs $cryptlib"
@@ -107,11 +107,11 @@ true)
 			linklibperl_nonshr='-lperl_nonshr'
 			;;
 		*)
-			shrpldflags="-H512 -T512 -bhalt:4 -bM:SRE -bE:perl.exp"
+			shrpldflags="-H512 -T512 -bhalt:4 -bE:perl.exp"
 			case "$osvers" in
-			3*)	shrpldflags="$shrpldflags -e _nostart"
+			3*)	shrpldflags="$shrpldflags -bM:SRE -e _nostart"
 				;;
-			*)	shrpldflags="$shrpldflags -b noentry"
+			*)	shrpldflags="$shrpldflags -G -bernotok -b noentry"
 				;;
 			esac
 			shrpldflags="$shrpldflags $ldflags $perllibs $cryptlib"
@@ -938,11 +938,19 @@ $(LIBPERL): $& $(perllib_dep) $(DYNALOADER) $(LIBPERLEXPORT)
 !NO!SUBS!
 		case "$osname" in
 		aix)
-			$spitshell >>$Makefile <<'!NO!SUBS!'
-	rm -f libperl$(OBJ_EXT)
-	mv $@ libperl$(OBJ_EXT)
-	$(AR) qv $(LIBPERL) libperl$(OBJ_EXT)
-!NO!SUBS!
+			bits=32
+			if test X"$use64bitall" = Xdefine; then bits=64; fi
+			$spitshell >>$Makefile <<!GROK!THIS!
+	rm -f shr.imp shr.o
+	mv \$@ shr.o
+	strip -e shr.o
+	( echo '#! libperl.so.${revision}.${patchlevel}(shr.o)' \\
+	; echo '# ${bits}' \\
+	; grep -v '^#!' perl.exp \\
+	) > shr.imp
+	\$(AR) qv \$@ shr.imp shr.o
+	rm -f shr.imp shr.o
+!GROK!THIS!
 			;;
 		esac
 		;;
-- 
2.27.0

